fix IOU = 0 case and add buffer

Rob1in opened the pull request:
Hungarian algorithm is always returning minimum but we need more check because if there is only one old detection and one new detection with `IOU=0`, it will be considered as a match. Making the PR now, but I need to test because I am removing a check

Rob1in commented on object_tracker/labels.go:
```diff
@@ -36,19 +36,19 @@ func (t *myTracker) RenameFromMatches(matches []int, oldDets, newDets []objdet.D
 	}
 	// Go through all NEW things and add them in (name them and start new track)
 	var freshDets []objdet.Detection
-	if len(newDets) > len(matches) {
```
@kharijarrett how bad is it to remove this check?

Rob1in commented:
It could also be something configurable, like:
 `if IOU<cfg.min_iou`

bhaney commented on object_tracker/labels.go:
```diff
@@ -36,19 +36,19 @@ func (t *myTracker) RenameFromMatches(matches []int, oldDets, newDets []objdet.D
-			newDets[newIdx] = ReplaceLabel(newDets[newIdx], oldDets[oldIdx].Label())
-			t.UpdateTrack(newDets[newIdx])
-			delete(notUsed, newIdx)
+			if matchinMtx[oldIdx][newIdx] != 0 {
```
Are there ways to check if the dim of matchinMtx would support these indices? Out of index accessing always causes a crash, so you want to check, or somehow ensure, that you won't accidently try to access something out of range

Rob1in replied:
this is true by "design" of the matching matrix

bhaney commented on object_tracker/labels.go:
```diff
@@ -36,19 +36,19 @@ func (t *myTracker) RenameFromMatches(matches []int, oldDets, newDets []objdet.D
-			t.UpdateTrack(newDets[newIdx])
-			delete(notUsed, newIdx)
+			if matchinMtx[oldIdx][newIdx] != 0 {
+				newDets[newIdx] = ReplaceLabel(newDets[newIdx], oldDets[oldIdx].Label())
```
same with accessing newDets and oldDets -- it would be safer to get the lengths of these slices and then do a quick comparison to make sure the indices fall within the lengths

Rob1in replied:
done

Rob1in commented:
Adding the buffer work. 

The test explains sequentially what happens. 

- In a nutshell, I added a buffer that stores lostDetections to run Hungarian Algorithm on what I called `allDetections`. 
- A lost detections is defined as a detection **within last detections**  that has no match in the new detections. 
- Before adding to the buffer we check if there is no other detection with the same unique label ID.
- If there, is oldest detection is deleted and the newest kept. See test for more. 
- New config attribute called `buffer_size`.
- Also changed how/what we store in an atomic

kharijarrett reviewed (approved):
LGTM! I think it's worth testing visually to see if there's any noticeable difference

bhaney reviewed (changes requested):
The new changes look all right, but I'd still like you to address the previous comments about checks if certain arrays will go out of range

bhaney reviewed (approved):
Thanks!
